### PR TITLE
Support creating smaller eStargz images (`--estargz-external-toc` and `--estargz-min-chunk-size`)

### PR DESCRIPTION
**Pull request by ktock**
This commit allows users optionally creating smaller eStargz images using the following flags to `ctr-remote i convert` or `ctr-remote i optimize`:

- `--estargz-external-toc`: Separates TOC JSON into another image (called "TOC image"). The result eStargz doesn't contain TOC so we can expect a smaller size than normal eStargz.

- `--estargz-min-chunk-size`: Specifies the minimal number of bytes of data must be written in one gzip stream. If it's > 0, multiple files and chunks can be written into one gzip stream. Smaller number of gzip header and smaller size of the result blob can be expected.

## About `--estargz-external-toc`

eStargz supports separating TOC into an external image called *TOC image*. This type of eStargz is the same as the normal eStargz but doesn't contain TOC JSON file (`stargz.index.json`) in the layer blob and has a special footer.

*TOC image* is an OCI image containing TOC. Each layer contains a TOC JSON file (`stargz.index.json`) in the root directory. Layer descriptors in the manifest of the TOC image must contain an annotation `containerd.io/snapshot/stargz/layer.digest`. The value of this is the digest of the eStargz layer corresponding to that TOC.

```json
{
  "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
  "digest": "sha256:64dedefd539280a5578c8b94bae6f7b4ebdbd12cb7a7df0770c4887a53d9af70",
  "size": 154425,
  "annotations": {
    "containerd.io/snapshot/stargz/layer.digest": "sha256:5da5601c1f2024c07f580c11b2eccf490cd499473883a113c376d64b9b10558f"
  }
}
```

Stargz snapshotter uses this annotation for searching TOC JSON of the mounting eStargz layer. Currently, it assumes the TOC image has the reference name same as the eStargz with `-esgztoc` suffix. For example, if an eStargz image is named `ghcr.io/stargz-containers/ubuntu:22.04-esgz`, stargz snapshotter acquires the TOC image from `ghcr.io/stargz-containers/ubuntu:22.04-esgz-esgztoc`. Note that future versions of stargz snapshotter will support more ways to search the TOC image (e.g. allowing custom suffix, using OCI Reference Type, etc.)

### Usage

convert:

```console
# ctr-remote i pull ghcr.io/stargz-containers/ubuntu:22.04
# ctr-remote i convert --oci --estargz --estargz-external-toc ghcr.io/stargz-containers/ubuntu:22.04 registry2:5000/ubuntu:22.04-ex
```

Layers in eStargz (`registry2:5000/ubuntu:22.04-ex`) don't contain TOC JSON. TOC image (`registry2:5000/ubuntu:22.04-ex-esgztoc`) contains TOC of all layers of the eStargz image.

Push them to the same registry:

```console
# ctr-remote i push --plain-http registry2:5000/ubuntu:22.04-ex
# ctr-remote i push --plain-http registry2:5000/ubuntu:22.04-ex-esgztoc
```

Pull it lazily:

```console
# ctr-remote i rpull --plain-http registry2:5000/ubuntu:22.04-ex
fetching sha256:14fb0ea2... application/vnd.oci.image.index.v1+json
fetching sha256:24471b45... application/vnd.oci.image.manifest.v1+json
fetching sha256:d2e4737e... application/vnd.oci.image.config.v1+json
# mount | grep "stargz on"
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/1/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
```

### Optional `--estargz-lossless` flag for lossless conversion

`ctr-remote i convert` supports an optional flag `--estargz-lossless` specified with `--estargz-external-toc`. This converts an image to eStargz without changing the diffID (uncompressed digest) so even eStargz-agnostic gzip decompressor (e.g. gunzip) can restore the original tar blob. `--estargz-record-in` can't be used with this flag.

## About `--estargz-min-chunk-size`

This option allows writing multiple small files into a single gzip stream. This can hopefully reduce the number of gzip headers and the size of the result blob.

To implement this, we need to introduce a new field `innerOffset` to TOC.

```golang
// TOCEntry is an entry in the stargz file's TOC (Table of Contents).
type TOCEntry struct {
...(omit)...
	// InnerOffset is an optional field indicates uncompressed offset
	// of this "reg" or "chunk" payload in a stream starts from Offset.
	// This field enables to put multiple "reg" or "chunk" payloads
	// in one chunk with having the same Offset but different InnerOffset.
	InnerOffset int64 `json:"innerOffset,omitempty"`
...(omit)...
```

This field allows the following structure.

minchunksize(1)

Unfortunatelly older version of stargz-snapshotter can't correctly understand the image created with `innerOffset` and can show a broken view of the filesystem.

### Usage

conversion:

```console
# ctr-remote i pull ghcr.io/stargz-containers/ubuntu:22.04
# ctr-remote i convert --oci --estargz --estargz-min-chunk-size=50000 ghcr.io/stargz-containers/ubuntu:22.04 registry2:5000/ubuntu:22.04-chunk50000
# ctr-remote i push --plain-http registry2:5000/ubuntu:22.04-chunk50000
```

Pull it lazily:

```console
# ctr-remote i rpull --plain-http registry2:5000/ubuntu:22.04-ex
fetching sha256:5d1409a2... application/vnd.oci.image.index.v1+json
fetching sha256:859e2b50... application/vnd.oci.image.manifest.v1+json
fetching sha256:c07a44b9... application/vnd.oci.image.config.v1+json
# mount | grep "stargz on"
stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/1/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
```

## Comparison

- Runner: GitHub Actions ubuntu:22.04 runner
  - Run: https://github.com/ktock/stargz-snapshotter/actions/runs/3290028313
  - Target commit: https://github.com/ktock/stargz-snapshotter/commit/13d388cd71a13bc47237b14f6650956ad197e9fe

### Comparison of image size and extraction time of a single file

This compares the size of images among different configurations. This also compares the avarage time to take for extracting a single file from the result blob. Note that we use the highest compression level of gzip (`gzip.BestCompression`) for eStargz by default. TOC image isn't included in the image size.

**[kdeneon/plasma:latest](https://github.com/orgs/stargz-containers/packages/container/kdeneon%2Fplasma/46138689?tag=latest)**

|benchmark|estargz/original (increased size)|average time to read file|
|---|---|---|
|min-chunk-size=0|1.022 (40866736 B)|5.83ms|
|min-chunk-size=0,externalTOC|1.019 (35536826 B)|5.78ms|
|min-chunk-size=1000|1.018 (33216221 B)|5.71ms|
|min-chunk-size=1000,externalTOC|1.015 (27881919 B)|5.77ms|
|min-chunk-size=10000|1.006 (12021539 B)|6.78ms|
|min-chunk-size=10000,externalTOC|1.004 (6663504 B)|6.75ms|
|min-chunk-size=25000|1.003 (6004067 B)|8.38ms|
|min-chunk-size=25000,externalTOC|1.000 (637889 B)|8.13ms|
|min-chunk-size=50000|1.002 (3857263 B)|10.7ms|
|min-chunk-size=50000,externalTOC|0.999 (-1506057 B)|10.6ms|

**[python:3.9-org](https://github.com/stargz-containers/image-ci/pkgs/container/python/7151254?tag=3.9-org)**

|benchmark|estargz/original (increased size)|average time to read file|
|---|---|---|
|min-chunk-size=0|1.033 (11350121 B)|4.74ms|
|min-chunk-size=0,externalTOC|1.029 (9933107 B)|4.74ms|
|min-chunk-size=1000|1.028 (9797255 B)|4.83ms|
|min-chunk-size=1000,externalTOC|1.024 (8378028 B)|4.72ms|
|min-chunk-size=10000|1.009 (3007542 B)|5.58ms|
|min-chunk-size=10000,externalTOC|1.005 (1582342 B)|5.57ms|
|min-chunk-size=25000|1.004 (1400329 B)|7.12ms|
|min-chunk-size=25000,externalTOC|1.000 (-26734 B)|7.04ms|
|min-chunk-size=50000|1.002 (807473 B)|9.91ms|
|min-chunk-size=50000,externalTOC|0.998 (-619813 B)|9.8ms|

**[ubuntu:22.04](https://github.com/orgs/stargz-containers/packages/container/ubuntu/40739314?tag=22.04)**

|benchmark|estargz/original (increased size)|average time to read file|
|---|---|---|
|min-chunk-size=0|1.066 (1996332 B)|5.01ms|
|min-chunk-size=0,externalTOC|1.061 (1841964 B)|4.99ms|
|min-chunk-size=1000|1.058 (1756194 B)|5.08ms|
|min-chunk-size=1000,externalTOC|1.053 (1601588 B)|4.99ms|
|min-chunk-size=10000|1.019 (589458 B)|5.83ms|
|min-chunk-size=10000,externalTOC|1.014 (434273 B)|5.76ms|
|min-chunk-size=25000|1.009 (279215 B)|7.52ms|
|min-chunk-size=25000,externalTOC|1.004 (123735 B)|7.37ms|
|min-chunk-size=50000|1.006 (169840 B)|10.1ms|
|min-chunk-size=50000,externalTOC|1.000 (14264 B)|10.2ms|

Image can be smaller using larger number to min-chunk-size and/or using exernal TOC. It's even possible to make eStargz smaller than the original image.

We need to be careful about that when we specify larger number to min-chunk-size, the time for extracting a file can be larger as well. One of the possible reasons is that files in one gzip stream isn't seekable so when we extract a file from the blob we also need to extract the neighboring files in the same stream.

To mitigate this performance drawback of min-chunk-size, stargz-snapshotter introduces a logic to aggressively and automatically cache files in a stream. When stargz-snapshotter extracts a file from one gzip stream and that stream contains multiple files, it automatically caches all of the neighboring files for speeding up accessing them in the future.

### Comparison of file read throughput (`tar`-ing the rootfs of the container)

This compares time to take `tar`-ing rootfs of the container running on stargz-snapshotter's FUSE.

Command: `/bin/bash -c 'time ( tar --exclude=/sys --exclude=/proc --exclude=/dev -cf - / | cat > /dev/null )'`

(average of 3 times)

|min-chunk-size|0|1000|10000|25000|50000|
|---|---|---|---|---|---|
|**[kdeneon/plasma:latest](https://github.com/orgs/stargz-containers/packages/container/kdeneon%2Fplasma/46138689?tag=latest)**|2m11.508s|2m11.162s|2m4.930s|2m3.928s|2m1.908s|
|**[python:3.9-org](https://github.com/stargz-containers/image-ci/pkgs/container/python/7151254?tag=3.9-org)**|29.451s|28.805s|27.241s|27.117s|25.928s|
|**[ubuntu:22.04](https://github.com/orgs/stargz-containers/packages/container/ubuntu/40739314?tag=22.04)**|3.986s|3.876s|3.580s|3.434s|3.368s|

No performance drawback has been observed for min-chunk-size-enabled images.

### HelloBench

- Runner: GitHub Actions ubuntu:22.04 runner
  - Run: https://github.com/ktock/stargz-snapshotter/actions/runs/3290028311
  - Target commit: https://github.com/ktock/stargz-snapshotter/commit/13d388cd71a13bc47237b14f6650956ad197e9fe

result

No performance drawback has been observed for min-chunk-size-enabled images.

## TODOs

- Allow configuring the name of TOC image (currently suffix `-esgztoc` is hardcoded)

### CODE REVIEW

File: docs/smaller-estargz.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+```console
+# ctr-remote i rpull --plain-http registry2:5000/ubuntu:22.04-ex
+fetching sha256:5d1409a2... application/vnd.oci.image.index.v1+json
+fetching sha256:859e2b50... application/vnd.oci.image.manifest.v1+json
+fetching sha256:c07a44b9... application/vnd.oci.image.config.v1+json
+# mount | grep "stargz on"
+stargz on /var/lib/containerd-stargz-grpc/snapshotter/snapshots/1/fs type fuse.rawBridge (rw,nodev,relatime,user_id=0,group_id=0,allow_other)
+```
+
+> NOTE: This flag creates an eStargz image with newly-added `innerOffset` funtionality of eStargz. Older version of Stargz Snapshotter cannot perform lazy pulling for the images created with this flag.
```
Comment (AkihiroSuda) [inline docs/smaller-estargz.md:79]: s/Older version/Versions before vX.Y.Z/

File: estargz/estargz.go [hunk 2]
```diff
@@ -207,8 +207,15 @@ func (r *Reader) initFields() error {
 	uname := map[int]string{}
 	gname := map[int]string{}
 	var lastRegEnt *TOCEntry
-	for _, ent := range r.toc.Entries {
+	var chunkTopIndex int
+	for i, ent := range r.toc.Entries {
 		ent.Name = cleanEntryName(ent.Name)
+		if ent.Type == "reg" || ent.Type == "chunk" {
```
Comment (AkihiroSuda) [inline estargz/estargz.go:213]: better to use `switch{}`, but can be another PR

File: estargz/estargz.go [hunk 3]
```diff
@@ -578,10 +602,48 @@ func (fr *fileReader) ReadAt(p []byte, off int64) (n int, err error) {
...
+		if !e.isDataType() {
+			continue
+		}
+		if e.Offset != fr.r.toc.Entries[ent.chunkTopIndex].Offset {
+			break
+		}
+		if in, err := io.CopyN(io.Discard, dr, e.InnerOffset-nr); err != nil || in != e.InnerOffset-nr {
+			return 0, fmt.Errorf("discard of remaining %d bytes != %v, %v", e.InnerOffset-nr, in, err)
+		}
+		nr += e.InnerOffset - nr
```
Comment (AkihiroSuda) [inline estargz/estargz.go:627]: Does this mean `nr = nr + e.InnerOffset - nr`, i.e., `nr = e.InnterOffset` ?

File: estargz/estargz.go [hunk 4]
```diff
@@ -599,11 +661,20 @@ type Writer struct {
...
+
 	// ChunkSize optionally controls the maximum number of bytes
 	// of data of a regular file that can be written in one gzip
 	// stream before a new gzip stream is started.
 	// Zero means to use a default, currently 4 MiB.
 	ChunkSize int
+
+	// MinChunkSize optionally controls the minimum number of bytes
+	// of data must be written in one gzip stream before a new gzip
+	// NOTE: This adds a TOC property that old reader doesn't understand.
```
Comment (AkihiroSuda) [inline estargz/estargz.go:674]: s/old/prior to vX.Y.Z/

File: estargz/externaltoc/externaltoc.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	gz, _ := gzip.NewWriterLevel(buf, gzip.NoCompression) // MUST be NoCompression to keep 51 bytes
+
+	// Extra header indicating the offset of TOCJSON
+	// https://tools.ietf.org/html/rfc1952#section-2.3.1.1
+	header := make([]byte, 4)
+	header[0], header[1] = 'S', 'G'
+	subfield := "STARGZEXTERNALTOC"                                   // len("STARGZEXTERNALTOC") = 17
+	binary.LittleEndian.PutUint16(header[2:4], uint16(len(subfield))) // little-endian per RFC1952
+	gz.Header.Extra = append(header, []byte(subfield)...)
+	gz.Close()
```
Comment (AkihiroSuda) [inline estargz/externaltoc/externaltoc.go:136]: catch err

File: metadata/testutil/testutil.go [hunk 6]
```diff
@@ -29,47 +30,32 @@ import (
...
-	*estargz.GzipDecompressor
-}
+var allowedPrefix = [4]string{"", "./", "/", "../"}
 
-func gzipCompressionWithLevel(compressionLevel int) compression {
-	return gzipCompression{estargz.NewGzipCompressorWithLevel(compressionLevel), &estargz.GzipDecompressor{}}
+var srcCompressions = map[string]tutil.CompressionFactory{
+	"zstd-fastest":                        tutil.ZstdCompressionWithLevel(zstd.SpeedFastest),
+	"zstd-default":                        tutil.ZstdCompressionWithLevel(zstd.SpeedDefault),
+	"zstd-bettercompression":              tutil.ZstdCompressionWithLevel(zstd.SpeedBetterCompression),
```
Comment (AkihiroSuda) [inline metadata/testutil/testutil.go:49]: Did you mean "zstd-speedbettercompression"?

File: metadata/testutil/testutil.go [hunk 7]
```diff
@@ -29,47 +30,32 @@ import (
...
+	"gzip-nocompression":                  tutil.GzipCompressionWithLevel(gzip.NoCompression),
```
Comment (AkihiroSuda) [inline metadata/testutil/testutil.go:50]: maybe:
```suggestion
	"gzip-no-compression":                  tutil.GzipCompressionWithLevel(gzip.NoCompression),
```

File: util/testutil/util.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+package testutil
+
+import (
+	"math/rand"
+)
+
+var runes = []rune("1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+
+func RandomContents(n int) string {
```
Comment (AkihiroSuda) [inline util/testutil/util.go:25]:
```suggestion
func RandomString(n int) string {
```

File: util/testutil/util.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = runes[rand.Intn(len(runes))]
+	}
```
Comment (AkihiroSuda) [inline util/testutil/util.go:29]: This would be more efficient https://pkg.go.dev/crypto/rand#Read

File: util/testutil/util.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+	return string(b)
+}
+
+func ViewContents(c []byte) string {
```
Comment (AkihiroSuda) [inline util/testutil/util.go:33]: What is this?

File: util/testutil/util.go (hunk 9 above)
Comment (AkihiroSuda) [reply util/testutil/util.go:29]: (Maybe use base64 for asciization)

File: cmd/ctr-remote/commands/convert.go [hunk 11]
```diff
@@ -69,6 +73,19 @@ When '--all-platforms' is given all images in a manifest list must be available.
...
+			Name:  "estargz-min-chunk-size",
+			Usage: "The minimal number of bytes of data must be written in one gzip stream. Note that this adds a TOC property that old reader doesn't understand.",
+			Value: 0,
+		},
+		cli.BoolFlag{
+			Name:  "estargz-external-toc",
+			Usage: "Separate TOC JSON into another image (called \"TOC image\"). The name of TOC image is the original + \"-esgztoc\" suffix. Both eStargz and the TOC image should be pushed to the same registry. stargz-snapshotter refers to the TOC image when it pulls the result eStargz image.",
+		},
+		cli.BoolFlag{
+			Name:  "estargz-lossless",
```
Comment (AkihiroSuda) [inline cmd/ctr-remote/commands/convert.go:86]: "Lossless images" may sound like graphic images as in PNG images 😆 

`estargz-keep-diff-id` might be less confusing, but no strong opinion

File: docs/estargz.md [hunk 12]
```diff
@@ -294,6 +310,60 @@ After the TOC is verified, the snapshotter mounts this layer using the metadata
 During runtime of the container, this snapshotter fetches chunks of regular file contents lazily.
 Before providing a chunk to the filesystem user, snapshotter recalculates the digest and checks it matches the one recorded in the corresponding TOCEntry.
 
+## eStargz image with an external TOC
+
+eStargz supports separating TOC into another image called *TOC image*.
+This type of eStargz is the same as the normal eStargz but doesn't contain TOC JSON file (`stargz.index.json`) in the layer blob and has a special footer.
```
Comment (AkihiroSuda) [inline docs/estargz.md:316]: It would be helpful to explain the motivation of this.

File: docs/smaller-estargz.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+```console
+# ctr-remote i pull ghcr.io/stargz-containers/ubuntu:22.04
+# ctr-remote i convert --oci --estargz --estargz-min-chunk-size=50000 ghcr.io/stargz-containers/ubuntu:22.04 registry2:5000/ubuntu:22.04-chunk50000
+# ctr-remote i push --plain-http registry2:5000/ubuntu:22.04-chunk50000
+```
+
+Pull it lazily:
+
+```console
+# ctr-remote i rpull --plain-http registry2:5000/ubuntu:22.04-ex
```
Comment (ktock) [inline docs/smaller-estargz.md:71]: typo:

```suggestion
# ctr-remote i rpull --plain-http registry2:5000/ubuntu:22.04-chunk50000
```

File: docs/smaller-estargz.md (hunk 1 above)
Comment (ktock) [reply docs/smaller-estargz.md]: Fixed

File: estargz/estargz.go (hunk 2 above)
Comment (ktock) [reply estargz/estargz.go]: Fixed to use `switch`.

File: estargz/estargz.go (hunk 3 above)
Comment (ktock) [reply estargz/estargz.go]: Yes. Refactored this to use `nr = e.InnterOffset` .

File: estargz/estargz.go (hunk 4 above)
Comment (ktock) [reply estargz/estargz.go]: Fixed

File: estargz/externaltoc/externaltoc.go (hunk 5 above)
Comment (ktock) [reply estargz/externaltoc/externaltoc.go]: Fixed

File: util/testutil/util.go (hunk 8 above)
Comment (ktock) [reply util/testutil/util.go]: Fixed

File: util/testutil/util.go (hunk 9 above)
Comment (ktock) [reply util/testutil/util.go:29]: Thank you for the suggestion. Fixed.

File: util/testutil/util.go (hunk 10 above)
Comment (ktock) [reply util/testutil/util.go]: This is a helper function to get a omitted string from a long string for printing. 
But the implementation looked wrong, so fixed.

File: cmd/ctr-remote/commands/convert.go (hunk 11 above)
Comment (ktock) [reply cmd/ctr-remote/commands/convert.go]: Thank you for the suggestion. Renamed to flag to `estargz-keep-diff-id`.

File: docs/estargz.md (hunk 12 above)
Comment (ktock) [reply docs/estargz.md:316]: Added the motivation.

File: docs/smaller-estargz.md (hunk 13 above)
Comment (ktock) [reply docs/smaller-estargz.md]: Fixed

File: util/testutil/util.go [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+	"testing"
+)
+
+// RandomString returns a random string of the specified number of bytes
+func RandomString(t *testing.T, n int) string {
+	b := make([]byte, n)
+	if _, err := rand.Read(b); err != nil {
+		t.Fatalf("failed rand.Read: %v", err)
+	}
+	return string(b)
```
Comment (AkihiroSuda) [inline util/testutil/util.go:30]: This can be non-ascii

Comment (ktock) [reply util/testutil/util.go]: Thank you for the review. I think it's fine that they're non-ascii. Fixed the function name to `RandomBytes`.

**Review by AkihiroSuda — APPROVED** [review]
